// ant-design-vue/lib/image/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "imageProps", {
  enumerable: true,
  get: function get() {
    return _Image.imageProps;
  }
});
Object.defineProperty(exports, "ImagePreviewGroup", {
  enumerable: true,
  get: function get() {
    return _PreviewGroup.default;
  }
});
exports.default = void 0;

var _vue = require("vue");

var _vcImage = _interopRequireDefault(require("../vc-image"));

var _Image = require("../vc-image/src/Image");

var _useConfigInject2 = _interopRequireDefault(require("../_util/hooks/useConfigInject"));

var _PreviewGroup = _interopRequireDefault(require("./PreviewGroup"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var Image = (0, _vue.defineComponent)({
  name: 'AImage',
  inheritAttrs: false,
  props: _Image.imageProps,
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;

    var _useConfigInject = (0, _useConfigInject2.default)('image', props),
        prefixCls = _useConfigInject.prefixCls;

    return function () {
      return (0, _vue.createVNode)(_vcImage.default, _extends(_extends(_extends({}, attrs), props), {
        prefixCls: prefixCls.value
      }), slots);
    };
  }
});
Image.PreviewGroup = _PreviewGroup.default;

Image.install = function (app) {
  app.component(Image.name, Image);
  app.component(Image.PreviewGroup.name, Image.PreviewGroup);
  return app;
};

var _default = Image;
exports.default = _default;
